records/2024-05-23.scalability.short-code-implementation.md

```diff
@@ -0,0 +1,107 @@
+
+```
+process_type is (bulk, normal, priority)
+```
+
+Given the new TemplateCategories, each template will default to a process_type that is set in TemplateCategory. To create manual overrides, an Admin will edit the process_type on the Templates Table. For this to work, we have two options:
+1. Create a new column in Templates `new_process_type`
+2. Clear the current process_type in the Templates table to be empty.
+
+After the above is implemented,
+```
+Notification Priority = Template.new_process_type else TemplateCategory.process_type
+```
+For most Templates, the Template.process_type/new_process_type will be empty. Only if overridden by Admin, will that field have information.
+
+**Use case:**  
+An SMS Template is of type "XYZ" that is of "low" priority. The client reaches out to us about the importance of the messages being sent immediately. We leave the category of the Template as "XYZ", but manually update the Template to "high" priority. This means that the message will be sent using the short-code.
+
+
+### Other Template Category
+
+Option 1:  
+Let client's manually add a Category for their template, if it doesn't fit in the above usecase. A ticket is created in Freshdesk for this new Template Category. A decision is made about adding a new Category, support can add that Category through an Admin workflow page (similar to Branding).
+
+Option 2:  
+Client adds a new Category. We run a classifier to the Templates and determine whether it should be a new Category, or live within the existing Categories (supervised learning). Depending on the decision, a new entry is made in the TemplateCategory table.
+
+Regardless of which option we choose, there is a time when the Category for a Template is undetermined. During this time, we will default the process_type of that Template to a Generic "Low Category (Bulk)" for all of our workflows to stay the same.
+
+#### Example
+There is a process_type in TemplateCategories and process_type in Templates.
+
+1. If a client chooses a category, the process_type in Templates will be empty.
+1. If the client chooses "Other" - while we are determining what TemplateCategory the Template must have, it will default to "Low".
+1. If an Admin wants to override the process_type for a given Template, that process_type will be overwritten in the Templates table.
+
+
+## Existing Templates
+
+In order to categorize existing templates, an offline classifier will be used. Currently Notify Production has 12600 existing templates. There are couple of methods that might be used to classify these templates. The size of the data file is not large, and we can assume that classification might be run locally.
+
+1. Classical Machine Learning Techniques - Naive Bayes, Linear Regression or Random Forest.  
+   We can create a test data set of around 100 templates, and use that to run any of the above models.
```

Comment:
Do you mean a training data set of around 100 templates? Not sure that'd be enough... and note that these methods would need the training data labelled (ie categories predetermined for the training data)

Reply:
The LLM option would definitely be a lot easier. We can get results out of the box using ChatGPT directly with proper prompt instructions.

Reply:
discussed irl